File: leet-code/19.删除链表的倒数第-n-个结点.js
/*
 * @lc app=leetcode.cn id=19 lang=javascript
 *
 * [19] 删除链表的倒数第 N 个结点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function (head, n) {
  let dummy = new ListNode(0, head)
  let fast = slow = dummy
  // 快指针先前进 n 步,此时 fast 指向删除节点
  while (n) {
    fast = fast.next;
    n--
  }

  // 让慢指针和快指针同步向前，
  // 因为要操作前一个节点，所以 fast 走到最后一个就可以了,不用走到null
  while (fast && fast.next) {
    fast = fast.next;
    slow = slow.next;
  }

  // 此时 slow 为删除的前置节点，slow.next 就是倒数第 n 个节点，删除它
  slow.next = slow.next.next;

  return dummy.next;
};
// @lc code=end

File: leet-code-rs/src/solutions/19_删除链表的倒数第-n-个结点.rs
/*
 * @lc app=leetcode.cn id=19 lang=rust
 *
 * [19] 删除链表的倒数第 N 个结点
 */
#[allow(unused)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}
// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
#[allow(unused)]
impl Solution {
    pub fn remove_nth_from_end(head: Option<Box<ListNode>>,  n: i32) -> Option<Box<ListNode>> {
        let dummy = ListNode { val: 0, next: head };
        let mut slow = &dummy;
        let mut fast = &dummy;

        // 右指针先向右走 n 步
        for _ in 0..n {
            fast = fast.next.as_ref()?;
        }

        while let Some(ref node) = fast.next {
            slow = slow.next.as_ref().unwrap();
            fast = node;
        }

        // 这里需要把 slow 从 &ListNode 强转成 &mut ListNode
        #[allow(mutable_transmutes)]
        let slow: &mut ListNode = unsafe { std::mem::transmute(slow) };
        slow.next = slow.next.take()?.next;

        return dummy.next;
    }
}
// @lc code=end

#[allow(unused)]
struct Solution;

#[cfg(test)]
mod test {
    #[allow(unused)]
    use super::*;

    #[test]
    fn test_remove_nth_from_end() {
        //assert_eq!(Solution::remove_nth_from_end(vec![]),[]);
        assert!(true)
    }
}
